api/provider/aws/service_create.go

```diff
@@ -1,8 +1,137 @@
 package aws
 
-import "github.com/quintilesims/layer0/common/models"
+import (
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/service/ecs"
+	"github.com/quintilesims/layer0/common/models"
+)
 
 func (s *ServiceProvider) Create(req models.CreateServiceRequest) (*models.Service, error) {
+	var (
+		cluster          string
+		desiredCount     int
+		loadBalancer     *ecs.LoadBalancer
+		loadBalancerRole string
+		serviceName      string
+		taskDefinition   string
+	)
+
+	fqEnvironmentID := addLayer0Prefix(s.Config.Instance(), req.EnvironmentID)
+	cluster = fqEnvironmentID
+
+	desiredCount = 1
+
+	serviceID := generateEntityID(req.ServiceName)
+	fqServiceID := addLayer0Prefix(s.Config.Instance(), serviceID)
+	serviceName = fqServiceID
+
+	taskDefinitionARN, err := s.lookupTaskDefinitionARNFromDeployID(req.DeployID)
+	if err != nil {
+		return nil, err
+	}
+
+	taskDefinition = taskDefinitionARN
+
+	if req.LoadBalancerID != "" {
+		fqLoadBalancerID := addLayer0Prefix(s.Config.Instance(), req.LoadBalancerID)
+		loadBalancerDescription, err := describeLoadBalancer(s.AWS.ELB, fqLoadBalancerID)
+		if err != nil {
+			return nil, err
+		}
+
+		taskDefinitionDescription, err := describeTaskDefinition(s.AWS.ECS, taskDefinitionARN)
+		if err != nil {
+			return nil, err
+		}
+
+		var loadBalancerContainer ecs.LoadBalancer
+		for _, containerDefinition := range taskDefinitionDescription.ContainerDefinitions {
+			for _, portMapping := range containerDefinition.PortMappings {
+				for _, listenerDescription := range loadBalancerDescription.ListenerDescriptions {
+					listener := listenerDescription.Listener
+					if *listener.InstancePort == *portMapping.ContainerPort {
+						loadBalancerContainer = ecs.LoadBalancer{
+							ContainerName:    containerDefinition.Name,
+							ContainerPort:    portMapping.ContainerPort,
+							LoadBalancerName: loadBalancerDescription.LoadBalancerName,
+						}
+					}
+				}
+			}
+		}
+
+		loadBalancer = &loadBalancerContainer
+		loadBalancerRole = fmt.Sprintf("%s-lb", fqLoadBalancerID)
+	}
+
+	if err := s.createService(desiredCount, cluster, serviceName, taskDefinition, loadBalancerRole, loadBalancer); err != nil {
+		return nil, err
+	}
+
+	if err := s.createTags(serviceID, req.ServiceName, req.EnvironmentID); err != nil {
+		return nil, err
+	}
+
+	return s.Read(serviceID)
+}
+
+func (s *ServiceProvider) createService(desiredCount int, cluster, serviceName, taskDefinition, loadBalancerRole string, loadBalancer *ecs.LoadBalancer) error {
```

Comment:
[nit] this order of params is kind of weird - one would expected cluster, serviceName, etc. to go first

Reply:
@zpatrick should we prefer
```cluster, servcieName, taskDefinition string, desiredCount int, loadBalancerRole string, loadBalancer *ecs.LoadBalancer
```
where `string` is specified twice,
```
cluster, serviceName, taskDefinition, loadBalancerRole string, desiredCount int, loadBalancer *ecs.LoadBalancer
```
where the `loadBalancer` and `loadBalancerRole` concepts are separated, or
```
cluster, serviceName, taskDefinition, loadBalancerRole string, loadBalancer *ecs.LoadBalancer, desiredCount int
```
where we don't maintain an increasing amount of complexity in parameters as we read left to right?

Reply:
I think the first reads the best

Reply:
Cool, thanks :+1:

✅